docs/cardano-monads.rst

```diff
@@ -0,0 +1,922 @@
+        type DBContext b
+        dbGet :: HasDBContext b ctx => DBTag -> ByteString -> ReaderT ctx b (Maybe ByteString)
+
+    class HasDBContext b ctx where
+        dbContext :: ctx -> DBContext b
+        -------- or: Lens' ctx (DBContext b)
+
+The concrete implementations are defined separately from the class,
+taking the context as an explicit parameter::
+
+    --       Constraints necessary for the impl (can contain 'MonadIO' if the impl is prod-specific,
+    --             \                             but preferably shouldn't)
+    --              \
+    --               \          NodeDBs are a RocksDB-specific context
+    --                \                  /
+    dbGetRocks :: MonadRealDB ctx m => NodeDBs -> DBTag -> ByteString -> ReaderT ctx m (Maybe ByteString)
+    dbGetRocks = ...
+
+    -- Constraints for the impl (no 'MonadIO', because it's used for tests)
+    --           \
+    --            \               DBPureVar is a pure-db specific context
+    --             \                /
+    dbGetPure :: MonadPureDB m => DBPureVar -> DBTag -> ByteString -> ReaderT ctx m (Maybe ByteString)
+    dbGetPure = ...
+
+We can use these implementations to define instances for the base monads::
+
+    instance MonadDBGet Prod where
+        type DBContext Prod = NodeDBs
+        dbGet tag key = do
+          nodeDBs <- asks dbContext
+          dbGetRocks nodeDBs tag key
+
+    instance MonadDBGet TestPure where
+        type DBContext TestPure = DBPureVar
+        dbGet tag key = do
+          dbPureVar <- asks dbContext
+          dbGetPure dbPureVar tag key
+
+These instances are defined not in the modules with implementations, but in the modules
+with the base monads (``Prod`` and ``TestPure``), similarly to what we have now.
+
+The implementation-specific context is extracted in these method definitions.
+
+In case of ``MonadDBGet``, the choice of implementation is determined by the
+base monad (in prod we use rocksdb, in tests we use pure db). However, we also
+have cases where we need runtime configurability, when with the same base monad
+there can be different implementations. For instance, for slotting we can have
+ntp-based slotting or simple slotting, depending on whether the ``--no-ntp`` flag
+was passed. In these cases we choose to use effect sums::
+
+    instance MonadSlotting Prod where
+        type SlottingContext Prod = Either SimpleSlottingVar NtpSlottingVar
+        getCurrentSlot = do
+          sctx <- asks slottingContext
+          either getCurrentSlotSimple getCurrentSlotNtp sctx
+
+While effect context sums are closed, we are fine with it because we don't
+need extensibility at instance definition site, while the implementations
+themselves are defined independently.
+
+Another example is passing ``SendActions``. With the new approach instead of
+``SendActions`` we should be passing ``SendContext`` (that contains the IP of
+the peer, etc), and define ``MonadSend`` to implement the actual sending logic.
```

Comment:
This is a red-flag for me. Is there some widely-held belief that the current solution regarding `SendActions m`, in which it's passed and used as a typical function parameter, is somehow problematic / not good enough?

Reply:
I sense a sort of general aversion to using functions, always favouring a `ReaderT` instead, but it's madness to me. It tends to make things *more* complex.

Reply:
The approach we ought to be taking is to aggressively favour simplicity. A solution such as this one, which demands two type classes and a type family for every capability / effect, had better have massive uncontroversial advantages over the simple record-passing solution, to pay for the added complexity.

Reply:
There are three major points against method records:

1. When one record needs to use another, it gets trapped in its closure, but then if you decide to pass another record, you can forget to update the closure. Mentioned it here: https://github.com/input-output-hk/cardano-sl/pull/1894#issuecomment-348498941
2. Lifting/hoisting method records is tricky, and we already have a precedent of doing it wrongly in our codebase (mentioned it in the document).
3. Method records allow too much freedom with regards to what can be put to them, it's hard to judge what code is being passed at any point of code (compare to classes, where all instances are global and there's a fixed set of them)

We can address (1) by avoiding using one method record from another, and rather than passing records for `A` and `B`, pass `A` and `A -> B` instead. This would mean constructing the actual record `B` each time we need to use its methods. So, far from perfect.

We can address (2) by never lifting and not using any monad transformers. This is also what we propose with the hybrid approach (forbid lifting), so here we seem to converge.

We can address (3) only by not using method records :man_shrugging: 

When we we discussing this with @georgeee I was a proponent of passing method records with taking the (1)-st point into account by using a special monad transformer which would be a fixed point of `ReaderT`. However, this fails the criteria of simplicity (which is your main selling point), and there's also point (3), so we abandoned this idea.

Reply:
>  which demands two type classes and a type family for every capability / effect

We would need one of the type classes (a `Has` class) even in case of passing method records explicitly, and for some method records we would need type families (the same way `Mockable` needs them), so it's not an unambiguous win for method records.

The actual advantage of method records is that since they allow arbitrary code to be passed, they are arbitrarily extensible, and there's no need for the effect context sums described in the doc. But that's a tradeoff, and it's not clear which way is better.

Reply:
> When one record needs to use another, it gets trapped in its closure, but then if you decide to pass another record, you can forget to update the closure.

If one record is used to implement another, then that's that, there's no reason to try to change it. That's the implementation and the things in its closure are part of that implementation. This is a good thing. One component `A :: Component1` can use one implementation of another component `B :: Component2`, without forcing every user of `A` to also use `B` as its `Component2` implementation.

> Lifting/hoisting method records is tricky, and we already have a precedent of doing it wrongly in our codebase (mentioned it in the document).

I don't think we'll need to. A record implementation which isn't fixed to one particular monad should say so in its type, by only constraining it rather than specializing it.

> Method records allow too much freedom with regards to what can be put to them, it's hard to judge what code is being passed at any point of code (compare to classes, where all instances are global and there's a fixed set of them)

This I don't understand. It's easy to judge what code is being passed: look at the definition of the record type. It's just as easy or very very slightly easier as with type classes.

```Haskell
readBlockRecord :: DB m -> HeaderHash -> m (Either DBError Block)
readBlockClass :: MonadReadBlock m => HeaderHash -> m (Either DBError Block)
```

Reply:
> If one record is used to implement another, then that's that, there's no reason to try to change it. That's the implementation and the things in its closure are part of that implementation. This is a good thing. 

I disagree. Let's say I want to disable logging locally: I can try to do it by passing `const ()` instead of a logging action (assuming that logging action is passed as a method record). But all the other method records have the real logging method in their closure, and it's hard to change.

Even the need to think about this is a disadvantage, it's hard to get this right.

> A record implementation which isn't fixed to one particular monad should say so in its type, by only constraining it rather than specializing it.

You propose to pass polymorphic records, like `MonadXXX m => Record m`? Well, this is asking for impredicative polymorphism now, or newtype wrappers, yet another inconvenience.

> This I don't understand. It's easy to judge what code is being passed: look at the definition of the record type.

With type classes the type determines the dictionary, with explicit method records nothing determines it, that's where the freedom comes from and its flip side (more difficult to reason).

Reply:
> We would need one of the type classes (a Has class) even in case of passing method records explicitly, and for some method records we would need type families (the same way Mockable needs them), so it's not an unambiguous win for method records.

No type class is necessary for record passing. Every term which needs the interface lamba-abstracts over the interface. We wouldn't go for some massive all-in-one record, we'd pass particular records for particular components.

> I disagree. Let's say I want to disable logging locally: I can try to do it by passing const () instead of a logging action (assuming that logging action is passed as a method record). But all the other method records have the real logging method in their closure, and it's hard to change.

Disabling logging locally means not disabling it in any antecedents (these things are not local, they're provided by the caller). What you're proposing is to disable logging globally, which can be done by providing a different logging interface everywhere such a thing is used.

> You propose to pass polymorphic records, like MonadXXX m => Record m? Well, this is asking for impredicative polymorphism now, or existential wrappers, yet another inconvenience.

No impredicative polymorphism nor existential wrappers necessary. Just constraints on the `m` parameter.

> The actual advantage of method records is that since they allow arbitrary code to be passed, they are arbitrarily extensible, and there's no need for the effect context sums described in the doc. But that's a tradeoff, and it's not clear which way is better (see point 3 in previous message).

Extensibility, sure. Do you have in mind some kind of global all-in-one record? That's not what I'd propose. I'm all for local record types for individual components. So yes in some sense they're very extensible because they're isolated. I can extend my DB component record independently of other components.

The actual advantage is something that you touch upon:

> With type classes the type determines the dictionary, with explicit method records nothing determines it

Yes! This is it, except that something does determine it: the explicit method records. They determine the type(s) which can carry them, whereas the type classes insist that the type itself determines the implementations (i.e. the records, implied by instances). With explicit method records, the implementation determines the monad. With type classes, the monad determines the implementation.

Both solutions effectively *do* the same thing: a record is passed along containing the particular implementations. The difference is that with the type class approach, in order to get a new implementation, you must invent a new type, and define a bunch of instances. These particular implementations sometimes depend upon *terms* rather than types, and so the type class approach demands that you pack up all of these antecedent terms into your `ReaderT` context and include that in the type itself.

This all strikes me as artificial; as trying very hard to make a type-level program where it doesn't naturally arise. We're *inventing* the monad which contains everything necessary to do what we need to do, rather than using an existing monad to express it, i.e. the base monad, which does all of the effects anyway, since reader is not an effect.

Drop the `ReaderT`, use the base monad, and pass the implementations that would be in the `ReaderT` context explicitly to where they're needed. Then we have the very same program, but with far fewer types and instances.

Reply:
> We wouldn't go for some massive all-in-one record, we'd pass particular records for particular components.

That would be impractical for more than one or two records, and we probably have dozens of them (just look at the size of constraint synonyms that we have now). The assumption is that we go with all-in-one parameter that contains individual records, because adding 10+ arguments to each function doesn't sound good.

> Disabling logging locally means not disabling it in any antecedents (these things are not local, they're provided by the caller). What you're proposing is to disable logging globally, which can be done by providing a different logging interface everywhere such a thing is used.

I have trouble understanding what you're saying here, my example is this:

```
f :: Logging m -> Networking m -> A -> m ()
g :: Logging m -> Networking m -> m ()

g logging networking = do
  f logging networking A1
  f logging networking A2
  f noLogging networking A3 -- oops
```

I tried to pass `noLogging` but the `networking` dictionary still uses `logging`. 

> The difference is that with the type class approach, in order to get a new implementation, you must invent a new type, and define a bunch of instances. 

Yes, that's a disadvantage, definitely. And it's asking for boilerplate. On the other hand, anyone can look at this type and know immediately what combination of records are passed everywhere it's used. These records are coherent (because type class instances are coherent). Whereas with explicit records we must deal with the fact that anyone can start passing different records at any point (as I demonstrated with `logging` / `noLogging` above), and it gets tricky. First, the compiler stops helping you with performance (no instance specialization), because it can no longer assume coherence. Second, the space of possible combinations of records explodes.

> These particular implementations sometimes depend upon terms rather than types

We propose to use effect context sums (described in the doc) to address this. Furthermore, there are effects where the choice of implementation *must* be reflected in the types. For instance, consider mutable references — they can be modeled as `IORef` in production and IO-enabled tests, but as `STRef` in pure tests. We must have a type family `type family Ref m`. Current `Mockable` machinery has these type families. It seems that there are effects for which we don't have a choice but to encode them in the types, so if we describe an approach based on method records, we need clear guidelines when to use a record and when to use a class.

> Drop the ReaderT, use the base monad, and pass the implementations that would be in the ReaderT context explicitly to where they're needed. Then we have the very same program, but with far fewer types and instances.

This _does_ sound compelling, as long as we figure out how to solve (1) the closure issue, (2) the "dozens of records" issue, (3) the inherently type-dependent effects issue.

So far the best solution for (1) and (2) that I see is a special monad `newtype EffectsT ctx m a = EffectsT (ctx (EffectsT ctx m) -> a)` (a fixed point of `ReaderT`). This seems exotic, over-complicated, and possibly won't even work. If you have better ideas I'd love to hear them. As for (3), it's less of a technical issue but more of a documentation issue (what's the decision process on when to use a record and when to use a class with a type family?). So let's focus on (1) and (2), these are the stumbling blocks that made me abandon the idea of method records.

Reply:
> I have trouble understanding what you're saying here, my example is this:
>
> ```
> f :: Logging m -> Networking m -> A -> m ()
> g :: Logging m -> Networking m -> m ()
> 
> g logging networking = do
>   f logging networking A1
>   f logging networking A2
>   f noLogging networking A3 -- oops
> ```
>
> I tried to pass noLogging but the networking dictionary still uses logging.

I'm saying that the inner workings for the `networking` dictionary are *not* local. Changing how/whether it logs is not a local change, it's a breach of abstraction. I'm also curious about how this would work in a `ReaderT ctx` solution.

```Haskell
f :: (HasLogging ctx m, HasNetworking ctx m) => A -> m ()
g :: (HasLogging ctx m, HasNetworking ctx m) => m ()

g = do
  -- Both of these use logging and networking determined by `m` (`ctx` presumably determined by `m` as well).
  f A1
  f A2
  -- Now to do `f` with no logging, we have to change `m` to `n`, where `n` has logging disabled.
  let noLogging = f A3 :: (HasLogging ctx n, HasNetworking ctx n) => n ()
  -- But we must get `n` back into `m`.
  -- What does it look like to do this?
```

> First, the compiler stops helping you with performance (no instance specialization), because it can no longer assume coherence.

I don't see how passing records would be less performant. Isn't what we have here equivalent to what goes on with a specialized instance? GHC replaces the instance use sites with record passing?

> Second, the space of possible combinations of records explodes — harder to reason. So it's more of a tradeoff than black/white.

The same "explosion" happens with instances, it's just an explosion at the type level instead.

>  there are effects where the choice of implementation must be reflected in the types. For instance, consider mutable references — they can be modeled as IORef in production and IO-enabled tests, but as STRef in pure tests. We must have a type family type family Ref m. Current Mockable machinery has these type families. It seems that there are effects for which we don't have a choice but to encode them in the types

These are legitimate effects *intrinsic* to the monad which carries them. `IO` doesn't need a term/record to inform how to create `IORef`s, so there's no problem here. What's at odds concerns "effects" like reading from a DB which are expressed entirely in terms of some existing base monad's effects, and an antecedent term.

> so if we describe an approach based on method records, we need clear guidelines when to use a record and when to use a class.

Use a record when the "effect" requires terms in order to carry it out. Use a class when the effect is intrinsic to the monad (it doesn't need any extra information). The `ReaderT ctx` solution works by constructing some monad which appears to express every desired effect intrinsically (the database effects are intrinsic, because the terms they require are built-in to the context).

Reply:
Ok I think I see how "local" effect changing would work: you'd take the reader context, and swap out one piece...

```Haskell
localEffect :: (ctx' -> ctx) -> ReaderT ctx m t -> ReaderT ctx' m t
localEffect = withReaderT
```

And the `ctx' -> ctx` could presumably be expressed by a lens?

Reply:
>  I'm also curious about how this would work in a ReaderT ctx solution.

To make sure we're on the same page, the idea is that we have the `ctx` decoupled from the base monad, so we wouldn't need `n ~> m` to change logging. With the proposed approach the code would look like this:

```
f :: (HasLogging ctx m, HasNetworking m) => A -> ReaderT ctx m ()
g :: (HasLogging ctx m, HasNetworking m) => m ()

g = do
  f A1
  f A2
  withReaderT (set loggingLens noLogging) $ do
    f A3
```

The reason that this is supposed not to have issues like with dictionaries is that `loggingLens` points not to some dictionary, but to something like `Maybe Handle`, and `noLogging` is expressed as `Nothing`. This way, when networking tries to access the `Maybe Handle` for logging, it will find `Nothing` and not log anything.

So it's not about `ReaderT` or explicit passing, it's about _what_ is being passed — monadic actions (where we have the issue with closures), or non-monadic data (like `Maybe Handle`).

> I don't see how passing records would be less performant. Isn't what we have here equivalent to what goes on with a specialized instance? GHC replaces the instance use sites with record passing?

I think the meaning of the term "specialization" is that instead of dictionary passing GHC creates a "specialized" version of the function (with a concrete dictionary), and this saves a pointer indirection to access methods. Record passing is the default strategy which is used when specialization doesn't fire (in case of polymorphic recursion).

> The same "explosion" happens with instances, it's just an explosion at the type level instead.

1. You could look at the module with instances and review this single module, whereas with records the methods can be passed/modified in arbitrary places in the code.
2. The "explosion" doesn't happen as long as the types are not nested. I.e. with a transformer per effect it's true that those transformers can be combined arbitrarily, causing the same explosion at the type level. But with a single base monad to do dispatching, the only combinations to review are the ones in the module with the instances.

Reply:
Anyway, let's pursue this idea of capabilities you expressed in the other thread, because it appears to solve the `noLogging` issue, unblocking the idea of using method records.

Reply:
These two approaches have the same expressive power.  With explicit dict passing you can implement networking to take logging as a parameter, and with `ReaderT` you can use `withReaderT` as above to alter the global config locally.  Adopting explicit dict passing is convenient in parts of the project (since it's very explicit).  There is the advantage of explicit dics, where you don't need the implicit parameters, because everything is done explicitly.